playground/backend/internal/db/datastore/emulator_wrapper.go

```diff
@@ -0,0 +1,182 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datastore
+
+import (
+	"beam.apache.org/playground/backend/internal/constants"
+	"beam.apache.org/playground/backend/internal/db/mapper"
+	"beam.apache.org/playground/backend/internal/logger"
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+)
+
+const (
+	emulatorHost    = "127.0.0.1"
+	portRangeStart  = 10000
+	portRangeEnd    = 20000
+	maxAttempts     = 3
+	startupDuration = 500 * time.Millisecond
+	pauseDuration   = 100 * time.Millisecond
+	waitDuration    = 10 * time.Second
+)
+
+type EmulatedDatastore struct {
+	*Datastore
+	emulator *emulator
+}
+
+type emulator struct {
+	Host string
+	Port int
+	cmd  *exec.Cmd
+}
+
+func (ed EmulatedDatastore) Close() error {
+	clientCloseErr := ed.Datastore.Client.Close()
+	emulatorStopErr := ed.emulator.Stop()
+	if clientCloseErr != nil {
+		return clientCloseErr
+	}
+	if emulatorStopErr != nil {
+		return emulatorStopErr
+	}
+	return nil
+}
+
+func NewEmulated(ctx context.Context) (EmulatedDatastore, error) {
+	emulator, err := startEmulator()
+	if err != nil {
+		return EmulatedDatastore{}, err
+	}
+
+	if _, ok := os.LookupEnv("GOOGLE_CLOUD_PROJECT"); ok {
+		err := os.Unsetenv("GOOGLE_CLOUD_PROJECT")
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if err := os.Setenv(constants.EmulatorHostKey, emulator.GetEndpoint()); err != nil {
+		panic(err)
+	}
+
+	datastoreDb, err := New(ctx, mapper.NewPrecompiledObjectMapper(), constants.EmulatorProjectId)
+	if err != nil {
+		return EmulatedDatastore{}, err
+	}
+
+	return EmulatedDatastore{Datastore: datastoreDb, emulator: emulator}, nil
+}
+
+func startEmulator() (*emulator, error) {
```

Comment:
It may make your code cleaner and easy to troubleshoot if you receive a `ctx context.Context` and use the pattern:
```
ctx, cancel := context.WithCancel(ctx)
defer cancel()
```

The test could perhaps do:
```
ctx, cancel := context.WithTimeout(...)
```

Would you like me to create an example?  Please let me know.

Reply:
If you I understand you correctly, you basically are suggesting to use context cancellation to kill the emulator process, right?
I've tried this approach, but sadly it doesn't work that well as cancelling the context only terminates the first process in the child process tree (starting the datastore emulator creates at least three processes), and the actual emulator process gets reparented by the `init`. The current code sends SIGTERM to the whole child process group to terminate the whole process tree.